Weekly_Work/Week1/Code_changes-git/Metrics_Root_class.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import json
+import datetime
+import matplotlib.pyplot as plt
+import pandas as pd
+
+
+# In[7]:
+
+
+class Metric():
+    
+    def __init__(self, path, since=None, until=None, source_code_exclude_list=None):
+
+        self.since = since
+        self.until = until
+        self.source_code_exclude_list = source_code_exclude_list
+        self.repo_urls = set()
+        
+        clean_commit_list = list()
+        clean_issue_list = list()
+        clean_pr_list = list()
+        
+        with open(path, 'r') as raw_data:
+            for line in raw_data:
+                line = json.loads(line)
+                self.repo_urls.add(line['origin'])
+                clean_line = dict()
+                if line['category'] == "commit":
+                    clean_line = self._clean_commit(line)
+
+                    if self._is_source_code(clean_line):
+                        clean_commit_list.append(clean_line)
+                    
+
+                elif line['category'] == "issue":
+                    if "pull_request" not in line['data']:
+                        clean_line = self._clean_issue(line)
+                        clean_issue_list.append(clean_line)
+                    else: continue
+                    
+
+                elif line['category'] == "pull_request":
+                    clean_line = self._clean_pr(line)
+                    clean_pr_list.append(clean_line)
+                        
+                        
+                self.clean_commit_df = pd.DataFrame(clean_commit_list)
+                self.clean_issue_df = pd.DataFrame(clean_issue_list)
+                self.clean_pr_df = pd.DataFrame(clean_pr_list)
+                
+            self.clean_dict = {
+                'commit': self.clean_commit_df,
+                'issue': self.clean_issue_df,
+                'pull_request': self.clean_pr_df
+            }
+       
+        if self.since:
+            for df in self.clean_dict.values():
+                df = df[df['created_date'] >= self._str_to_dt_other(self.since)]
+        else: 
+            self.since = self._get_date("since")
+            
+        if self.until:
+            for df in self.clean_dict.values():
+                df = df[df['created_date'] < self._str_to_dt_other(self.until)]
+        else: 
+            self.until = self._get_date("until")
+        
+    def _clean_commit(self, line):
```

Comment:
What is this for? Commits should come as commits in the JSON file...

Reply:
I am not completely sure what you are referring to here, but it is probably the `_clean_commit` function. This method is for cleaning the raw data line present in the Perceval data.
A "Perceval" commit (or for that matter, an issue or a pull request) contains several attributes which may not be of our concern, like the links to the commit's repository or maybe the uuid. This function (and likewise, `_clean_pr` and `_clean_issue`) does exactly that.

Reply:
At least for a first implementation, I would rather use the items produced by perceval as they are, when we're dealing with lists. We can find out later if it makes sense to remove some fields or something. For sure that will make a more efficient implementation, but would also make it a bit more difficult to understand. I suggest that we start easy, and then we can find out if the complexity added by a certain optimization makes sense.

In the case of dataframes, we need a flat structure (data for a row), and that would mean we need to extract the fundamental fields from the dataframe. That's one of the reasons why I think we need separate implementations for each of the data sources (commits, prs, issues) in this case, with the construction of the dataframe done at the root level, but by calling a function on each item that will be implemented by child classes.